Simplify report replay checking. (#574)
ietf-wg-ppm/draft-ietf-ppm-dap: draft-ietf-ppm-dap.md

Report Share:
: An encrypted input share comprising a piece of a report.

Time Window:
: An interval of time associated with a measurement task, whose duration is
  equal to the task's time precision and whose start has a UNIX timestamp value

>   which is a multiple of the task's time precision.

Reviewer: I might also mention briefly what this is used for (replay protection for one!)

| batchQueriedMultipleTimes  | A batch was queried with multiple distinct aggregation parameters. |
| batchMismatch              | Aggregators disagree on the report shares that were aggregated in a batch. |
| unauthorizedRequest        | Authentication of an HTTP request failed (see {{request-authentication}}). |

> | stepMismatch               | The Aggregators disagree on the current step of the DAP aggregation sub-protocol. |

Reviewer 2: Orthongonal nit: back when we got a review from HTTPDIR, we got a recommendation to use the term "interaction" or "sub-interaction" instead of "sub-protocol". Mostly this was cleaned up in #450, but I appear to have missed this one.

Note that the `leader_aggregator_url` and `helper_aggregator_url` values may
include arbitrary path components.

In addition, in order to facilitate the aggregation and collection

> sub-protocols, each of the Aggregators is configured with following parameters:

Reviewer 2: As above: prefer "interaction" to "sub-protocol", per HTTPDIR's advice.

      number generator.

    * `time` is the time at which the report was generated. The Client SHOULD

>       truncate this value down to the nearest multiple of the task's

Reviewer 2: I don't object to using the verb "truncate" here (is the precedent [Go's `time.Truncate`](https://pkg.go.dev/time#Time.Truncate)?), but "truncate [...] down" is awkward. There's no such thing as "truncating up" so why say "truncate down"?
```suggestion
      truncate this value to the nearest multiple of the task's
```

the implementation note in {{input-share-validation}}.

The Leader MUST ignore any report pertaining to a batch that has already been
collected (see {{input-share-validation}} for details). Otherwise, comparing the
aggregate result to the previous aggregate result may result in a privacy
violation. Note that this is also enforced by the Helper during the aggregation

> interaction. The Leader MAY also abort the upload sub-protocol and alert the

Reviewer 2: As above: prefer "interaction" to "sub-protocol", per HTTPDIR's advice.

Client with error `reportRejected`.

The Leader MAY ignore any report whose timestamp is past the task's

> `task_expiration`. When it does so, it SHOULD also abort the upload sub-protocol

Reviewer 2: As above: prefer "interaction" to "sub-protocol", per HTTPDIR's advice.

and alert the Client with error `reportRejected`. Client MAY choose to opt out
of the task if its own clock has passed `task_expiration`.

The Leader may need to buffer reports while waiting to aggregate them (e.g.,
while waiting for an aggregation parameter from the Collector; see
{{collect-flow}}). The Leader SHOULD NOT accept reports whose timestamps are too
far in the future. Implementors MAY provide for some small leeway, usually no
more than a few minutes, to account for clock skew. If the Leader rejects a

> report for this reason, it SHOULD abort the upload sub-protocol and alert the

Reviewer 2: As above: prefer "interaction" to "sub-protocol", per HTTPDIR's advice.

guaranteed to combine into a valid, refined measurement.

VDAF preparation is mapped onto an aggregation job as illustrated in

> {{agg-flow}}. The aggregation sub-protocol is comprised of a sequence of HTTP

Reviewer 2: As above: prefer "interaction" to "sub-protocol", per HTTPDIR's advice.

In general, reports cannot be aggregated until the Collector specifies an
aggregation parameter. However, in some situations it is possible to begin
aggregation before a collection request arrives. For example, Prio3 has just one
valid aggregation parameter (the empty string); or, perhaps the relevant
aggregation parameter can be determined in a deployment-specific way. In these

> cases, aggregation can begin as soon as the Leader is confident that no

Reviewer 3: Maybe be explicit about the requirement instead "aggregation can begin as soon as the leader is willing to discard any further incoming reports for this time window."

opaque AggregationJobID[16];
~~~

> The set of candidate reports MUST produce a set of report ID ranges which does

Reviewer 3: I agree that calling the new thing a "report ID range" makes everything hard to read. This sentence looks like it is comparing report ID ranges but it is actually comparing "report ID ranges".

Considering the fact that you're saying "a set of" here. Does that mean the old behavior would still be allowed by making each range have exactly one report ID?

Reviewer 3: To clarify, I am wondering why anybody would want to do this given that the report IDs are randomly assigned. Is the intent to allow people to recover from some amount of late arriving reports?

Author: > I agree that calling the new thing a "report ID range" makes everything hard to read. This sentence looks like it is comparing report ID ranges but it is actually comparing "report ID ranges".

Indeed; I am careful to use the term "interval" where an "interval" is truly needed. But "report ID range" does not capture that a "report ID range" includes a time window, too.

> Considering the fact that you're saying "a set of" here. Does that mean the old behavior would still be allowed by making each range have exactly one report ID?

Not quite -- an aggregation job may include reports in multiple time windows. In this case, there would be one "report ID range" for each time window represented in the aggregation job.

> To clarify, I am wondering why anybody would want to do this given that the report IDs are randomly assigned. Is the intent to allow people to recover from some amount of late arriving reports?

Nope, I think late-arriving reports are effectively lost (with high probability) in this scheme. The "report ID ranges" in use are all "maximal" in the sense that they stretch from the smallest to the largest available report ID.

Reviewer 3: I think if the "set" only covers the "different time intervals" part you need to clarify. The way it's written the ranges don't need to be maximal. Somebody could send 3 ranges for t1 and 7 ranges for t2 as longs as they in total cover all the reports. And in the limit, send one range per report and everything turns into linear search again.

Author: I think the ranges do need to be maximal; there actually isn't much wiggle room here. 

The relevant Leader text is:

> The set of candidate reports MUST produce a set of report ID commitments which
does not conflict with the report ID commitments for any other aggregation job
associated with the task (see {{report-id-commitments}}).

[note that "report ID range" got renamed to "report ID commitment" here]

The relevant Helper text is:

> 1. Check if the report has been previously aggregated. If so, the input share
   MUST be marked as invalid with the error `report_replayed`. A report is
   considered previously aggregated if it conflicts with a report ID commitment
   for a completed aggregation job (see {{report-id-commitments}}).

That is, both fall back to the report-id-commitments section; this section specifies a maximal range:

> The report ID commitments associated with a given set of reports are determined by the following procedure:
> 
> 1. Partition the set of reports into groups based on the time window their
>    client timestamp falls into; this might be achieved by partitioning the
>    reports on a key generated by truncating each report's timestamp down to the
>    nearest multiple of the task's `time_precision`.
> 2. For each partitioned group of reports, generate an interval from the smallest
>    report ID in the group to the largest report ID in the group, inclusive. The
>    report ID commitment for this partition is this interval, associated with the
>    time window that generated this partition.

I had previously believed there was some wiggle room for what the Helper commits if the Leader fails in its `MUST` & sends an aggregation job containing a report which falls into an existing report ID commitment. However, now I think the Helper's behavior is fully specified here with no wiggle room:

* Any reports conflicting with an existing report ID commitment must be marked as `report_replayed`.
* The new commitment associated with the aggregation job is still the maximal report ID commitments suggested by the report-id-commitments section -- this commitment will conflict with an already-existing commitment, but that's OK. (A clever implementation might merge conflicting commitments without changing its externally-observable behavior, but this isn't required.)

Author: By the way -- I'm interested in your thoughts on the new text in the report-id-commitments section @simon-friedberger. Other than renaming "report ID ranges" to "report ID commitments", I tried to make it clearer that the set of report ID commitments is one-per-time-window, rather than allowing the aggregators to generate a larger count of smaller commitment intervals.

> The set of candidate reports MUST produce a set of report ID ranges which does

Reviewer: How strict do you think we should be? As we discussed on slack today, at a minimum we want the helper to be able to reject any range that overlaps with an already aggregated range. But the helper could make a smarter decision by "merging" the two ranges and rejecting only the overlap, in which case the MUST here would not be enforced by the helper.

Author: I think a `MUST` is appropriate here for the Leader: the Leader must follow this suggestion to avoid having otherwise-acceptable reports rejected as a replay.

The Helper can indeed make a few different choices about how to deal with an aggregation job which does intersect with a previous aggregation job. The relevant text for the Helper is currently:

> Check if the report has been previously aggregated. If so, the input share MUST be marked as invalid with the error `report_replayed`. A report is considered previously aggregated if it intersects with a report ID range for a completed aggregation job (see {{report-id-ranges}}).

... which implies to me that we have to effectively allow "merging" of ranges if they intersect, in the case that an intersecting aggregation job is received. (A Helper which does something more specific, e.g. marking individual report IDs as used, would work to avoid replays; but I'm not sure a Leader could take advantage of this without knowing the specific strategy used by the Helper.)

I'm personally happy with this level of flexibility, but maybe the Helper should be more be more flexible? If so, we might consider relaxing the Leader's MUST here.

Reviewer: I prefer MUST here as well.

If all of the above checks succeed, the input share is not marked as invalid.

#### Report ID ranges {#report-id-ranges}

Replay checking is implemented using "report ID ranges". A report ID range is an
interval of report IDs (using lexicographic ordering), along with an associated
timestamp.

A set of reports for a task determines a set of report ID ranges. The report ID
ranges associated with a given set of reports are determined by the following
procedure:

1. Round each report's timestamp down to the nearest multiple of
   `time_precision`, and partition the set of reports into groups based on this

>    rounded timestamp.

Reviewer: Let's make sure this is harmonized with the rest of the draft. I believe we say "truncate", not "round". However I haven't checked.

Author: We actually use both! I prefer `truncate` so I switched to that.

If all of the above checks succeed, the input share is not marked as invalid.

#### Report ID ranges {#report-id-ranges}

> Replay checking is implemented using "report ID ranges". A report ID range is an

Reviewer: ```suggestion
Replay protection is implemented using "report ID ranges". A report ID range is an
```

If all of the above checks succeed, the input share is not marked as invalid.

#### Report ID commitments {#report-id-commitments}

Replay protection is implemented using "report ID commitments". A report ID

> commitment is an interval of report IDs (using lexicographic ordering),

Reviewer 2: An interval of time makes sense, but an interval of report IDs is a strange notion.
```suggestion
commitment is an ordered set of report IDs (using lexicographic ordering),
```

associated with a given time window (that is, an interval of time whose duration
is equal to the task's `time_precision` and whose start has a UNIX timestamp
which is a multiple of the task's `time_precision`).

A set of reports for a task determines a set of report ID commitments, producing
one commitment per time window represented in the set of reports. The report ID
commitments associated with a given set of reports are determined by the
following procedure:

1. Partition the set of reports into groups based on the time window their
   client timestamp falls into; this might be achieved by partitioning the

>    reports on a key generated by truncating each report's timestamp down to the

Reviewer 2: ```suggestion
   reports on a key generated by truncating each report's timestamp to the
```

`AggregationJobContinueReq` messages contain a `step` field, allowing
Aggregators to ensure that their peer is on an expected step of the DAP

> aggregation sub-protocol. In particular, the intent is to allow recovery from a

Reviewer 2: As above: prefer "interaction" to "sub-protocol", per HTTPDIR's advice.

- Support the aggregate interaction, which includes validating and aggregating
  reports; and
- Publish and manage an HPKE configuration that can be used for the upload

>   sub-protocol; and

Reviewer 2: As above: prefer "interaction" to "sub-protocol", per HTTPDIR's advice.

- Implement and store state for the replay mitigation in
  {{input-share-validation}}.

Beyond the minimal capabilities required of Helpers, Leaders are generally
required to:

> - Support the upload sub-protocol and store reports; and

Reviewer 2: As above: prefer "interaction" to "sub-protocol", per HTTPDIR's advice.

- Track batch report size during each collect flow and request encrypted output
  shares from Helpers.

### Collector Capabilities

Collectors statefully interact with Aggregators to produce an aggregate output.
Their input to the protocol is the task parameters, configured out of band,
which include the corresponding batch window and size. For each collection
invocation, Collectors are required to keep state from the start of the

> sub-protocol to the end as needed to produce the final aggregate output.

Reviewer 2: As above: prefer "interaction" to "sub-protocol", per HTTPDIR's advice.

Collectors must also maintain state for the lifetime of each task, which
includes key material associated with the HPKE key configuration.

> ## Aggregation Jobs & Replay Protection

Reviewer: Note that we never use "&" in place of "and" in this draft.
```suggestion
## Aggregation Jobs and Replay Protection
```

Author: aagh, sorry, I keep making that mistake

### Leader

> As part of the aggregation protocol, the Leader is responsible for generating

Reviewer: nit: consistency. I'd be cool with saying "protocol" instead of "sub-protocol", but let's track in an issue.
```suggestion
As part of the aggregation sub-protocol, the Leader is responsible for generating
```

Author: 👍🏻 It turned out there were a bunch of places that got this wrong -- I fixed them all up. There were a few places where it was ambiguous; for the most part I made as few changes as possible while still referring to "sub-protocols" when discussing only the upload/aggregation/collection portions of DAP.

aggregation jobs. The Leader must produce aggregation jobs using reports with

> non-overlapping intervals of report IDs for each time window (see {{agg-init}}).

Reviewer: nit: consistency. FWIW I weakly prefer "overlap" to "intersect".
```suggestion
non-intersecting intervals of report IDs for each time window (see {{agg-init}}).
```

> non-overlapping intervals of report IDs for each time window (see {{agg-init}}).

Reviewer: Let's make sure "time window" is defined in the terms section of the doc

In general, the Aggregators are required to keep state for tasks and all valid
reports for as long as collection requests can be made for them. However, it is
not necessary to store the complete reports. Each Aggregator only needs to store
an aggregate share for each batch, along with a flag indicating whether the

> aggregate share has been collected. This is due to the requirement that in the

Reviewer: I believe the old text still applies: a batch interval is a sequence of time windows; and the batch interval is not known at aggregation time.

Author: Hmm, I think the old text was slightly inaccurate for the time-interval case: aggregators don't have to store an aggregate share for each possible a batch interval (i.e., sequence of time windows) -- an aggregator needs to store an aggregate share for each _time window_.

I had felt that distinguishing between the batch modes was not so useful here; I restored the text, modified for accuracy.

share has been collected. This is due to the requirement that in the
an aggregate share for each batch, along with a flag indicating whether the
aggregate share has been collected. This is due to the requirement that in the
time-interval case, the batch interval respect the boundaries defined by the DAP
parameters; and that in leader-selected case, a batch is determined entirely by
a batch ID. (See {{batch-validation}}.)

However, Aggregators are also required to implement several per-report checks
that require retaining a number of data artifacts. For example, to detect replay
attacks, it is necessary for each Aggregator to retain the set of report IDs of
reports that have been aggregated for the task so far. Depending on the task
lifetime and report upload rate, this can result in high storage costs. To
alleviate this burden, DAP allows Aggregators to drop this state as needed, so
long as reports are dropped properly as described in {{input-share-validation}}.
Aggregators SHOULD take steps to mitigate the risk of dropping reports (e.g., by

> evicting the oldest data first).

Reviewer: I think the core idea here still applies: it's just that we're storing report ranges rather than IDs.

Author: OK -- I dropped this because we'll now be storing perhaps a few thousand ID ranges per time window, but no objection to keeping this text around.

Reviewer: I'm just suggesting we update the text.

An important parameter of a DAP deployment is the minimum batch size. If a batch
includes too few reports, then the aggregate result can reveal information about
individual measurements. Aggregators enforce the agreed-upon minimum batch size

> during the collection sub-protocol, but implementations SHOULD also opt out of

Reviewer 2: As above: prefer "interaction" to "sub-protocol", per HTTPDIR's advice.
